src/client/activation/migrateDataScienceSettingsService.ts

```diff
@@ -72,13 +79,38 @@ export class MigrateDataScienceSettingsService implements IExtensionActivationSe
         });
 
         await this.fs.writeLocalFile(filePath, fileContents);
-        return fileContents;
+    }
+
+    // Users may have mapped old python.dataScience.* commands to custom keybindings
+    // in their user keybindings.json. Ensure we migrate these too.
+    private async fixKeybindingsFile(filePath: string) {
+        const fileContents = await this.fs.readLocalFile(filePath);
+        const errors: ParseError[] = [];
+        const keybindings = parse(fileContents, errors, { allowTrailingComma: true, disallowComments: false });
+        if (errors.length > 0) {
+            traceError('JSONC parser returned ParseError codes', errors);
+            return;
+        }
+        if (!Array.isArray(keybindings)) {
+            return;
+        }
+        keybindings.forEach((keybinding) => {
+            if (typeof keybinding.command === 'string') {
+                keybinding.command = keybinding.command.replace('python.dataScience.', 'jupyter.');
+            }
+        });
+        const migratedKeybindings = JSON.stringify(keybindings, undefined, 4);
+        await this.fs.writeLocalFile(filePath, migratedKeybindings);
     }
 
     @traceDecorators.error('Failed to update test settings')
     private async updateSettings(resource: Resource): Promise<void> {
-        const filesToBeFixed = await this.getFilesToBeFixed(resource);
-        await Promise.all(filesToBeFixed.map((file) => this.fixSettingInFile(file)));
+        const filesToBeFixed = (await this.getFilesToBeFixed(resource)).map((file) => this.fixSettingsFile(file));
+        const userCustomKeybindingsFile = this.application.userCustomKeybindingsFile;
+        if (userCustomKeybindingsFile && (await this.fs.localFileExists(userCustomKeybindingsFile))) {
+            filesToBeFixed.push(this.fixKeybindingsFile(userCustomKeybindingsFile));
```

Comment:
Will this fix the files for users who have already been upgraded?
Or is this code going to work only for users who have not ugraded to Jupyter extension?

Reply:
Today this runs on every extension activation, so yes it will fix for users who have already upgraded. But I think we need to add some global state to ensure the migration happens once from here onward.

Reply:
> But I think we need to add some global state to ensure the migration happens once from here onward.

Yes we should.